sympy/sets/sets.py

```diff
@@ -670,33 +677,51 @@ class ProductSet(Set):
     is_ProductSet = True
 
     def __new__(cls, *sets, **assumptions):
-        def flatten(arg):
-            if isinstance(arg, Set):
-                if arg.is_ProductSet:
-                    return sum(map(flatten, arg.args), [])
-                else:
-                    return [arg]
-            elif iterable(arg):
-                return sum(map(flatten, arg), [])
-            raise TypeError("Input must be Sets or iterables of Sets")
-        sets = flatten(list(sets))
+        if len(sets) == 1 and iterable(sets[0]) and not isinstance(sets[0], (Set, set)):
+            SymPyDeprecationWarning(
+                feature="ProductSet(iterable)",
+                useinstead="ProductSet(*iterable)",
+                issue=17557,
+                deprecated_since_version="1.5"
+            ).warn()
+            sets = tuple(sets[0])
 
-        if EmptySet() in sets or len(sets) == 0:
-            return EmptySet()
+        sets = [sympify(s) for s in sets]
 
-        if len(sets) == 1:
-            return sets[0]
+        if not all(isinstance(s, Set) for s in sets):
+            raise TypeError("Arguments to ProductSet should be of type Set")
+
+        # Nullary product of sets is *not* the empty set
+        if len(sets) == 0:
+            return FiniteSet(())
+
+        if EmptySet() in sets:
+            return EmptySet()
 
         return Basic.__new__(cls, *sets, **assumptions)
 
+    @property
+    def sets(self):
+        return self.args
+
+    def flatten(self):
+        def _flatten(sets):
+            for s in sets:
+                if s.is_ProductSet:
+                    for s2 in _flatten(s.sets):
+                        yield s2
+                else:
+                    yield s
+        return ProductSet(*_flatten(self.sets))
+
     def _eval_Eq(self, other):
         if not other.is_ProductSet:
             return
 
-        if len(self.args) != len(other.args):
+        if len(self.sets) != len(other.sets):
             return false
 
-        return And(*(Eq(x, y) for x, y in zip(self.args, other.args)))
+        return And(*(Eq(x, y) for x, y in zip(self.sets, other.sets)))
```

Comment:
I think that it will likely give a error when reaching this line because of the issue in #17574, 
but I also think that it would be better to return an unevaluated equality in one level like ![image](https://user-images.githubusercontent.com/34944973/64211484-4e695480-cee1-11e9-98f0-ee668cb1d8f9.png) rather than returning an expanded form like ![image](https://user-images.githubusercontent.com/34944973/64211510-66d96f00-cee1-11e9-987c-739f5baae0f5.png)

Reply:
That makes sense. BTW feel free to push changes on here if you are using this as a base.

Reply:
I've pushed this change now